bigqueryNullFloat64 add +Inf, -Inf and NaN support

deankarn opened the pull request:
The following is a PR to fix issue https://github.com/googleapis/google-cloud-go/issues/2575 to enable support for `+Inf`, `-Inf` and `Nan` which BigQuery supports but Go's json package does not.

This is related to PR https://github.com/googleapis/google-cloud-go/pull/2408 but implements the fix slightly differently supporting short form `+Inf` and `-Inf` so less bytes have to go across the wire.

deankarn commented:
@googlebot I signed it!

deankarn commented:
@googlebot I signed it!

deankarn commented:
@shollyman I corrected the original commit not having the proper commit email and ensured the email is also registered with my GitHub account. Not sure why it's still saying CLA is not signed, perhaps needs to be force ran again?

Please let me know if anything else is required that I've missed so I can correct it :)

deankarn commented on bigquery/nulls.go:
```diff
@@ -181,6 +196,12 @@ func (n *NullFloat64) UnmarshalJSON(b []byte) error {
 	n.Float64 = 0
 	if bytes.Equal(b, jsonNull) {
 		return nil
+	} else if bytes.HasPrefix(b, posInf) || bytes.HasPrefix(b, posInf2) {
```
Hey @shollyman 

I agree that `HasPrefix` might lead to corner cases and makes sense to change that to equality checks instead :) will do that shortly.

I was also using `HasPrefix` for two other reasons:
1. When Go itself prints values for `Infinity`, `-Infinity` it does so using the short form `+Inf` and `-Inf` (which the BigQuery api accepts happily)
2. The source of some data I'm parsing was also using the short form that was produced by Go.

To avoid a heap of customization in my, and others, code base' I was hoping to keep the unmarshaling short form support.

What do you think about :
- Changing this to equality checks only, no `HasPrefix` to avoid any corned cases.
- Change to always Marshal the long form of these values `Infinity`, `-Infinity` and `NaN`
- Keep the Unmarshal support for the short form equality checks as well as the long form. Doing this will avoid people having to create a custom type to support the way Go itself, and other languages, output this kind of data.

I can't see any harm also supporting the short form **for unmarshaling purposes only**, thoughts?

shollyman replied:
That seems reasonable.  It's definitely the case that BQ is following the robustness principle in terms of parsing the variation, so it's reasonable we include that when unmarshalling.

deankarn replied:
Thanks will make the changes and ping you in the PR once done.

deankarn commented:
@shollyman I updated the PR based on our conversation and added dedicated tests for `NullFloat64`

please let me know if there's anything else I can do to get this merged :)

shollyman commented:
One last item:  can you edit your PR to allow edits by maintainers?  Neither I nor our automation bots can update the branch to merge it currently.

deankarn commented:
@shollyman I had to create a new PR because GitHub doesn't allow setting edits by maintainers on organization forks, only user forks. (may want to mention this in your contributing guidelines)

Will close this one and the new PR is here just merged them to my user fork https://github.com/googleapis/google-cloud-go/pull/2618

buyology commented:
@shollyman can you please explain why this gets reviewed and merged and not #2408? 

I first submitted this change a year ago (https://code-review.googlesource.com/c/gocloud/+/44750) and even then left with no review — and was the first to submit anything when you migrated to GitHub.

Feels pretty demotivating to come up with any further changes if they just get ignored!

shollyman commented:
@buyology My apologies for overlooking your PR (twice, sadly in this case).  I recently modified how I consume github notifications, hence my attention on a newer PR.  If I've overlooked other feedback/contributions from you, feel free to reach out via email (same username @google.com) to discuss.
